drift/core/beamtransfer.py

```diff
@@ -831,190 +831,260 @@ def _generate_svdfiles(self, regen=False, skip_svd_inv=False):
+                        beam = np.dot(ut3, bfr)
 
-        # If we're part of an MPI run, synchronise here.
-        mpiutil.barrier()
+                        # Save out the evecs (for transforming from the telescope frame into the SVD basis)
+                        dset_ut[fi, :nmodes] = ut * noisew[np.newaxis, :]
 
-        # Collect the spectrum into a single file.
-        self._collect_svd_spectrum()
+                        # Save out the modified beam matrix (for mapping from the sky into the SVD basis)
+                        dset_bsvd[fi, :nmodes] = beam.reshape(
+                            nmodes, self.telescope.num_pol_sky, self.telescope.lmax + 1
+                        )
+
+                        if not skip_svd_inv:
+                            # Find the pseudo-inverse of the beam matrix and save to disk.
+                            # First try la.pinv, which uses a least-squares solver.
+                            try:
+                                ibeam = la.pinv(beam)
+                            except la.LinAlgError as e:
+                                # If la.pinv fails, try la.pinv2, which is SVD-based and
+                                # more likely to succeed. If successful, add file attribute
+                                # indicating pinv2 was used for this frequency.
+                                print(
+                                    "***pinv failure: m = %d, fi = %d. Trying pinv2..."
+                                    % (mi, fi)
+                                )
+                                try:
+                                    ibeam = la.pinv2(beam)
+                                    if "inv_bsvd_from_pinv2" not in fs.attrs.keys():
+                                        fs.attrs["inv_bsvd_from_pinv2"] = [fi]
+                                    else:
+                                        bad_freqs = fs.attrs["inv_bsvd_from_pinv2"]
+                                        fs.attrs[
+                                            "inv_bsvd_from_pinv2"
+                                        ] = bad_freqs.append(fi)
+                                except:
+                                    # If pinv2 fails, print error message
+                                    raise Exception(
+                                        "pinv2 failure: m = %d, fi = %d" % (mi, fi)
+                                    )
+
+                            dset_ibsvd[fi, :, :, :nmodes] = ibeam.reshape(
+                                self.telescope.num_pol_sky,
+                                self.telescope.lmax + 1,
+                                nmodes,
+                            )
+
+                        # Save out the singular values for each block
+                        dset_sig[fi, :nmodes] = sig
+
+                # Write a few useful attributes.
+                fs.attrs["baselines"] = self.telescope.baselines
+                fs.attrs["m"] = mi
```

Comment:
might be redundant since it's part of the file name..

Reply:
True - I am just copying what was in the code already